contracts/poap/src/contract.rs

```diff
@@ -0,0 +1,451 @@
+    let creator = deps.api.addr_validate(&msg.event_info.creator)?;
+
+    // Check that start time is before the end time
+    if !msg.event_info.end_time.gt(&msg.event_info.start_time) {
+        return Err(ContractError::StartTimeAfterEndTime {
+            start: msg.event_info.start_time,
+            end: msg.event_info.end_time,
+        });
+    }
+
+    // Check that the start time is in the future
+    if !msg.event_info.start_time.gt(&env.block.time) {
+        return Err(ContractError::StartTimeBeforeCurrentTime {
+            current_time: env.block.time,
+            start_time: msg.event_info.start_time,
+        });
+    }
+
+    // Check that the end time is in the future
+    if !msg.event_info.end_time.gt(&env.block.time) {
+        return Err(ContractError::EndTimeBeforeCurrentTime {
+            current_time: env.block.time,
+            end_time: msg.event_info.end_time,
+        });
+    }
+
+    // Check that the poap uri is a valid IPFS url
+    let poap_uri = Url::parse(&msg.event_info.base_poap_uri)
+        .map_err(|_err| ContractError::InvalidPoapUri {})?;
+    if poap_uri.scheme() != "ipfs" {
+        return Err(ContractError::InvalidPoapUri {});
+    }
+
+    // Check that the event uri is a valid IPFS url
+    let event_uri =
+        Url::parse(&msg.event_info.event_uri).map_err(|_err| ContractError::InvalidEventUri {})?;
+    if event_uri.scheme() != "ipfs" {
+        return Err(ContractError::InvalidEventUri {});
+    }
+
+    // Check per address limit
+    if msg.event_info.per_address_limit == 0 {
+        return Err(ContractError::InvalidPerAddressLimit {});
+    }
+
+    let config = Config {
+        admin: admin.clone(),
+        minter: minter.clone(),
+        per_address_limit: msg.event_info.per_address_limit,
+        cw721_code_id: msg.cw721_code_id,
+        mint_enabled: false,
+    };
+    // Save the received event info.
+    CONFIG.save(deps.storage, &config)?;
+
+    let event_info = EventInfo {
+        creator: creator.clone(),
+        start_time: msg.event_info.start_time,
+        end_time: msg.event_info.end_time,
+        base_poap_uri: msg.event_info.base_poap_uri.clone(),
+        event_uri: msg.event_info.event_uri.clone(),
+    };
+    // Save the event info
+    EVENT_INFO.save(deps.storage, &event_info)?;
+
+    // Submessage to instantiate cw721 contract
+    let sub_msgs: Vec<SubMsg> = vec![SubMsg {
+        msg: WasmMsg::Instantiate {
+            admin: Some(admin.to_string()),
+            code_id: msg.cw721_code_id,
+            msg: to_binary(&Cw721InstantiateMsg {
+                name: msg.cw721_initiate_msg.name,
+                symbol: msg.cw721_initiate_msg.symbol,
+                minter: env.contract.address.to_string(),
+            })?,
+            funds: info.funds,
+            label: "poap cw721".to_string(),
+        }
+        .into(),
+        id: INSTANTIATE_CW721_REPLY_ID,
+        gas_limit: None,
+        reply_on: ReplyOn::Success,
+    }];
```

Comment:
It can be simplified to:
```rust
SubMsg::reply_on_success(
    wasm_instantiate(msg.cw721_code_id, &msg.cw721_initiate_msg, info.funds, "poap_cw721".into())?,
    INSTANTIATE_CW721_REPLY_ID,
)
```

Reference: 
[submessage](https://github.com/CosmWasm/cosmwasm/blob/main/packages/std/src/results/submessages.rs#L55)
[wasm_instantiate](https://github.com/CosmWasm/cosmwasm/blob/9bf920b00b76f31f3c73ed3632cd07cbd6a94b7b/packages/std/src/results/cosmos_msg.rs#L201)